accessible/workspace_treeview.component.js

```diff
@@ -0,0 +1,234 @@
+        string = string + ' ' + arguments[i];
+      }
+      return string;
+    },
+    hasNoPreviousConnection: function(block){
+      if (!block.previousConnection){
+        return 'blockly-disabled';
+      } else {
+        return undefined;
+      }
+    },
+    hasNoNextConnection: function(block){
+      if (!block.nextConnection){
+        return 'blockly-disabled';
+      } else {
+        return undefined;
+      }
+    },
+    checkParentList: function(parentList) {
+      console.log('setting parent list');
+      var tree = parentList;
+      while (tree && !tree.id.match(/^blockly-workspace-tree\d+$/)) {
+        tree = tree.parentNode;
+      }
+      if (tree && tree.getAttribute('aria-activedescendant') == parentList.id){
+        this.treeService.updateSelectedNode(parentList, tree, false);
+      }
+    },
+    addClass: function(node, classText) {
+      //ensure that node doesn't have class already in it
+      var classList = node.className;
+      classList = classList.split(' ');
+      var canAdd = true;
+      for (var i=0; i<classList.length; i++){
+        if (classList[i] == classText) {
+          canAdd = false;
+        }
+      }
+      //add class if it doesn't
+      if (canAdd) {
+        if (classList.length == 0) {
+          node.className += classText;
+        } else {
+          node.className += (' ' + classText);
+        }
+      }
+    },
+    setId: function(block){
+      if (this.isTopBlock){
+        return this.parentId+'-node0';
+      }
+      return this.treeService.createId(block);
+    },
+    getInfo: function(block) {
+      //List all inputs
+      if (this.infoBlocks[block.id]) {
+        //TODO(madeeha): is there a situation in which overwriting often unnecessarily is a problem?
+        this.infoBlocks[block.id].length = 0;
+      } else {
+        this.infoBlocks[block.id] = [];
+      }
+
+      var blockInfoList = this.infoBlocks[block.id];
+
+      for (var j = 0, field; field = block.fieldRow[j]; j++) {
+        blockInfoList.push(field);
+      }
+
+      return this.infoBlocks[block.id];
+    },
+    inputType: function(connection) {
+      if (connection.check_) {
+        return connection.check_.join(', ').toUpperCase();
+      } else {
+        return 'any';
+      }
+    },
+    sendToSelected: function(block){
+      if (this.sharedClipboardService) {
+        this.sharedClipboardService.pasteToMarkedConnection(block);
+        block.dispose(true);
+        alert('block sent to marked spot');
+      }
+    },
+    inputMenuSelected: function(connection, event) {
+      switch (event.target.value) {
+        case 'MARK_SPOT':
+          this.sharedClipboardService.markConnection(connection);
+          console.log('marked spot');
+          break;
+        case 'PASTE':
+          this.sharedClipboardService.paste(connection);
+          break;
+      }
+      event.target.selectedIndex = 0;
+    },
+    notCompatibleWithClipboard: function(connection) {
```

Comment:
This function is being used to determine if a button should be enabled or disabled. In html, in order for a button to be enabled, the html should read <button disabled=undefined> and in order for it to be disabled, the button should read <button disabled=[any value]>. Therefore, this function cannot return false, but *must* return undefined instead. I also would rather not change the name to isCompatibleWithClipboard because it will still need to return undefined if the connection is compatible with what's on the clipboard, which is unintuitive. Let me know if I should change the comments on line 208 and 211 to make that more clear.

I currently have the function returning 'blockly-disabled' if the function should be disabled so that I can use notCompatibleWithClipboard in the setLabelledBy function call on line 49. While it maybe isn't the most conventional decision, I feel that it is cleaner than having yet another function that calls notCompatibleWithClipboard and simply returns 'undefined' or 'blockly-disabled'.